### DISCUSSION TITLE
Preferred method for large, centred maps?

### DISCUSSION THREAD

**Discussion by BrettWitty** [StarArawn/bevy_ecs_tilemap#174] [Q&A] [answered]
At the core of it, `TilePos` is a pair of `u32`s. Since it's unsigned and starts at zero, we've implicitly anchored our map with (0,0). With a fixed map, this is okay, but it feels awkward if you want, say, a procedural map that radiates from a central point.

`bevy_tilemap` uses `i32` (and a 3-dim tile position) so it's easy to anchor a map at (0,0) and radiate outwards.

In `bevy_ecs_tilemap` we could start our map at (2^31,2^31) and be roughly equivalent, but then the numbers are gross, and you hope that the library doesn't default-instantiate the gigantic empty space. I believe it's clever enough not to, but I didn't find clear guidance in the docs.

A similar question applies for z-levels where you need to know how many layers you'll want below the default z-layer, since in  `bevy_ecs_tilemap` it's a `u16`.

What's the blessed approach to centered maps like this?

**Comment by StarArawn** [comment, marked as answer]
Hi!

Ideally for large maps you should chunk. This allows you to use u64 or even u128 for global tile positions. `TilePos` must be a unsigned integer.  This is because of how tiles are stored in an array (Negative numbers cannot exist in an array) You can use a hash map for tile storage but there are performance issues when accessing tiles within a hash map. If you would like centered tile maps I would suggest using a transform to shift the map visually instead of having negative tile coordinates.

A structure that would likely work out great for this might look something like this:

Chunk:
```rust
// Use a bundle here and wrap data in Components.
pub struct Chunk {
  pub tile_storage: Vec<Option<Entity>>,
  pub chunk_pos: IVec2,
  pub transform: Transform, 
}
```

Tile:
```rust
// Use a bundle here and wrap data in Components.
pub struct Tile {
  pub texture_id: u16,
  pub local_tile_pos: UVec2,
  pub global_tile_pos: (i64, i64),
};
```

With the new API design this is doable. See:
https://github.com/StarArawn/bevy_ecs_tilemap/blob/rewrite/src/tiles/mod.rs#L54
https://github.com/StarArawn/bevy_ecs_tilemap/blob/rewrite/src/tiles/storage.rs

I'm planning on creating an infinite map example in the coming future which will showcase some of this off. 👍
